### PR TITLE
Allow same subnet traffic

### LINKED ISSUE #3320: Allow same subnet traffic with subnet ACL
### Feature request

The user can decide whether the subnet with  ACL  allows the same subnet traffic through the configuration parameter

### PR DESCRIPTION
**Pull request by lynn901**
- [ ] Make sure you have followed [Kube-OVN Code Style](https://github.com/kubeovn/kube-ovn/blob/master/CODE_STYLE.md).

### What type of this PR
Examples of user facing changes:
- Features

### Which issue(s) this PR fixes:
Fixes #3320 

### <samp>🤖 Generated by Copilot at 51c27ec</samp>

This pull request adds support for creating and updating ACL rules for logical switches with different CIDR blocks and subnet traffic policies. It also enables TLS verification for the OVSDB client. It modifies the `SubnetSpec` type, the `UpdateLogicalSwitchACL` method and its mock and test functions, and the `handleAddOrUpdateSubnet` function. It also adds a new constant `AllowSameSubnetPriority` and changes the `InsecureSkipVerify` field of the `tlsConfig` struct.

### <samp>🤖 Generated by Copilot at 51c27ec</samp>

> _Sing, O Muse, of the mighty deeds of the OVN developers,_
> _Who with skill and wisdom modified the code of the SubnetSpec,_
> _Adding a new field to allow or block the traffic of the same subnet,_
> _And changing the TLS config to secure the connection to the OVSDB._

### <samp>🤖 Generated by Copilot at 51c27ec</samp>

*  Add `AllowSameSubnetTraffic` field to `SubnetSpec` type to configure ACL rules for logical switch ([link](https://github.com/kubeovn/kube-ovn/pull/3321/files?diff=unified&w=0#diff-69f75d53361877cc02ab6f413f306486d29287c6400b92b71c4c74f018f3234aL161-R162))
*  Modify `UpdateLogicalSwitchACL` method of `ACL` and `MockACL` interfaces to accept `cidrBlock` and `allowSameSubnetTraffic` parameters ([link](https://github.com/kubeovn/kube-ovn/pull/3321/files?diff=unified&w=0#diff-2eadf7fb3dd75cdfd04457a9b7e244a2bfb9aa9bb790ff6ed89b86599c0a4c71L1329-R1331), [link](https://github.com/kubeovn/kube-ovn/pull/3321/files?diff=unified&w=0#diff-2eadf7fb3dd75cdfd04457a9b7e244a2bfb9aa9bb790ff6ed89b86599c0a4c71L1337-R1339), [link](https://github.com/kubeovn/kube-ovn/pull/3321/files?diff=unified&w=0#diff-50fe35ba11ad04edb53bfaccef915df3650a81844495ed73d3c158e975c04c49L110-R110))
*  Modify `UpdateLogicalSwitchACL` method of `OVNNbClient` and `MockNbClient` interfaces to match `ACL` interface and pass parameters to underlying `MockACL` or `OVNNbClient` methods ([link](https://github.com/kubeovn/kube-ovn/pull/3321/files?diff=unified&w=0#diff-2eadf7fb3dd75cdfd04457a9b7e244a2bfb9aa9bb790ff6ed89b86599c0a4c71L3589-R3591), [link](https://github.com/kubeovn/kube-ovn/pull/3321/files?diff=unified&w=0#diff-2eadf7fb3dd75cdfd04457a9b7e244a2bfb9aa9bb790ff6ed89b86599c0a4c71L3597-R3599))
*  Modify `UpdateLogicalSwitchACL` method of `OVNNbClient` struct in `pkg/ovs/ovn-nb-acl.go` to create or update ACL rules for logical switch based on `cidrBlock` and `allowSameSubnetTraffic` parameters ([link](https://github.com/kubeovn/kube-ovn/pull/3321/files?diff=unified&w=0#diff-cebcd3f05c126053ae717f2f1e1fae98b5b1112936636b0d6ee8ba860932b7c6L420-R420), [link](https://github.com/kubeovn/kube-ovn/pull/3321/files?diff=unified&w=0#diff-cebcd3f05c126053ae717f2f1e1fae98b5b1112936636b0d6ee8ba860932b7c6L428-R428), [link](https://github.com/kubeovn/kube-ovn/pull/3321/files?diff=unified&w=0#diff-cebcd3f05c126053ae717f2f1e1fae98b5b1112936636b0d6ee8ba860932b7c6R437-R461))
*  Modify `handleAddOrUpdateSubnet` function in `pkg/controller/subnet.go` to pass `CIDRBlock` and `AllowSameSubnetTraffic` fields of subnet resource to `UpdateLogicalSwitchACL` method of `OVNNbClient` interface ([link](https://github.com/kubeovn/kube-ovn/pull/3321/files?diff=unified&w=0#diff-3742843fd11c637b26eae183fb8a0c4eeff96bbfb18a8db1e0ded0c33733f205L829-R829))
*  Modify `testUpdateLogicalSwitchACL` function in `pkg/ovs/ovn-nb-acl_test.go` to test `UpdateLogicalSwitchACL` method with different `cidrBlock` and `allowSameSubnetTraffic` values and check expected ACL rules ([link](https://github.com/kubeovn/kube-ovn/pull/3321/files?diff=unified&w=0#diff-5fec517e195bd3aa2f06602f2c19f83df3351068b887f6ffbc5710ff974d96b8R658), [link](https://github.com/kubeovn/kube-ovn/pull/3321/files?diff=unified&w=0#diff-5fec517e195bd3aa2f06602f2c19f83df3351068b887f6ffbc5710ff974d96b8L677-R678), [link](https://github.com/kubeovn/kube-ovn/pull/3321/files?diff=unified&w=0#diff-5fec517e195bd3aa2f06602f2c19f83df3351068b887f6ffbc5710ff974d96b8R684-R699))
*  Modify `NewOvsDbClient` function in `pkg/ovsdb/client/client.go` to set `InsecureSkipVerify` field of `tlsConfig` struct to false, to enable verification of server's certificate and hostname when using TLS ([link](https://github.com/kubeovn/kube-ovn/pull/3321/files?diff=unified&w=0#diff-8c46f500ad7d8fe51f1165ee87bbfe97e665d1856bd02101a2b85cc0ea9946adL78-R78))
*  Add `AllowSameSubnetPriority` constant to `util` package in `pkg/util/const.go` to assign high priority to ACL rules that allow traffic within same subnet ([link](https://github.com/kubeovn/kube-ovn/pull/3321/files?diff=unified&w=0#diff-23fe1930ce750ec27bfa79a8fb8d076245c64df2b62ea6d6367cdd59260b164cR143-R144))

### CODE REVIEW

File: pkg/ovsdb/client/client.go [hunk 1]
```diff
@@ -75,7 +75,7 @@ func NewOvsDbClient(db, addr string, dbModel model.ClientDBModel, monitors []cli
 		tlsConfig := &tls.Config{
 			Certificates:       []tls.Certificate{cert},
 			RootCAs:            certPool,
-			InsecureSkipVerify: true,
+			InsecureSkipVerify: false,
```
Comment (zhangzujian) [inline pkg/ovsdb/client/client.go:78]: Why change this value?

Comment (bobz965) [reply pkg/ovsdb/client/client.go:78]: oh，he changes this because of a lint err. will remove later. 

the `// #nosec` before this line has no effect.

Comment (lynn901) [reply pkg/ovsdb/client/client.go:78]: I will checkout this change

**Review by bobz965 — APPROVED** [review]

**Comment by zhangzujian** [comment]
CRD should be updated.

**Comment by bobz965** [comment]
hi @lynn901 is there any doc about this? please attach it

**Comment by lynn901** [comment]
> CRD should be updated.

ok

File: pkg/ovs/ovn-nb-acl_test.go [hunk 2]
```diff
@@ -655,6 +655,7 @@ func (suite *OvnClientTestSuite) testUpdateLogicalSwitchACL() {
 
 	ovnClient := suite.ovnClient
 	lsName := "test_update_acl_ls"
+	cidrBlock := "192.168.2.0/24, 2409:8720:4a00::0/64"
```
Comment (zhangzujian) [inline pkg/ovs/ovn-nb-acl_test.go:658]:
```suggestion
	cidrBlock := "192.168.2.0/24,2409:8720:4a00::0/64"
```

**Comment by zhangzujian** [comment]
`AllowSameSubnetTraffic` is kind of confusing. We should make the field name clear.

**Comment by lynn901** [comment]
> `AllowSameSubnetTraffic` is kind of confusing. We should make the field name clear.

can we use allowSubnetInternalTrafficWithACL or allowSubnetInternalTraffic?

**Pull request closed without merging**